Changed UserInterface with identity + improved unit tests and doc

ezimuel opened the pull request:
This PR changes the `getUsername` of `UserInterface` in `getIdentity()`, as suggested in #14.

This PR contains also the following update:
- improvement on error management of `UserRepository\PdoDatabase::authenticate` and `UserRepository\PdoDatabase::getRolesFromUser`;
- improvement of the doc with the `UserRepository` adapter configuration values;
- improvement of the unit test with the Exception throw for the new error management of `UserRepository\PdoDatabase`.

**Note:** since this PR will change UserInterface it's a BC break. It must be released in 0.3. Moreover, this change will affect some `zend-expressive-authentication` adapters, like `zend-expressive-authentication-session`. We need to refactor these adapters as well.

ezimuel committed:
Refactored UserInterface with identity + improved unit tests and doc

ezimuel committed:
Updated the CHANGELOG

ezimuel committed:
Removed exception + improved tests

ezimuel committed:
Fixed CS issue

michalbundyra commented on test/UserRepository/PdoDatabaseTest.php:
```diff
@@ -20,32 +20,44 @@ public function testConstructor()
+    }
+
+    /**
+     * @expectedException Zend\Expressive\Authentication\Exception\RuntimeException
```
I would suggest to use method `expectException` just before line where the exception should be thrown, otherwise it could be thrown in any line of this test...

ezimuel replied:
Good catch, thanks!

michalbundyra commented on test/UserRepository/PdoDatabaseTest.php:
```diff
@@ -69,36 +75,63 @@ public function testAuthenticateInvalidUsername()
     }
+
+    /**
+     * @expectedException Zend\Expressive\Authentication\Exception\RuntimeException
```
as above, please use method just before line where the exception should be thrown

michalbundyra commented on test/UserRepository/PdoDatabaseTest.php:
```diff
@@ -69,36 +75,63 @@ public function testAuthenticateInvalidUsername()
+    }
+
+    /**
+     * @expectedException Zend\Expressive\Authentication\Exception\InvalidConfigException
```
the same

michalbundyra commented on test/UserRepository/PdoDatabaseTest.php:
```diff
@@ -20,32 +20,44 @@ public function testConstructor()
-                'username' => 'username',
-                'password' => 'password',
+                'identity' => 'username',
+                'password' => 'password'
```
please re-add trailing comma

weierophinney commented on CHANGELOG.md:
```diff
@@ -2,15 +2,16 @@
 
-- Nothing
+- [#14](https://github.com/zendframework/zend-expressive-authentication/issues/14)
+  changed the `getUsername` with `getIdentity` in UserInterface.
```
Change to:

```markdown
renames the method `UserInterface::getUsername()` to `UserInterface::getIdentity()`.
```

Depending on another suggested change below, we may also need to detail something about the return value change.

weierophinney commented on docs/book/v1/intro.md:
```diff
@@ -22,11 +22,11 @@ namespace Zend\Expressive\Authentication;
      * @return string
      */
-    public function getUsername(): string;
+    public function getIdentity(): string;
```
My understanding of #14 is that the reporter was also requesting that this method could return more than just a string - an object, for instance? If that's the case, we'd also need to remove the return type here, and modify the `@return` annotation to reflect that.

That said, I'd prefer we just have it return a string, and then the individual `UserInterface` implementations could provide other details via additional properties or methods.

geerteltink replied:
> Get the unique user identity (id, username, email address or ...)

With _id_ I meant an integer or an UUID even. But I guess using `return (string) $this->id;` would solve the return type.

ezimuel replied:
@weierophinney and @xtreamwayz ok, so we'll stay with a `string` return.

michalbundyra replied:
We are using single space before colon in return type declaration (consistency)

weierophinney commented on src/UserRepository/PdoDatabase.php:
```diff
@@ -41,14 +42,19 @@ public function __construct(PDO $pdo, array $config)
-        $stmt->bindParam(':username', $credential);
+        if (false === $stmt) {
+            throw new Exception\RuntimeException(
+                "Error during the user authentication, please check the configuration"
```
I'd change this to "An error occurred when preparing to fetch user details from the repository; please verify your configuration." This will give more information to the developer, without giving security details out should the exception message leak to a consumer.

ezimuel replied:
@weierophinney good catch, thanks!

ezimuel committed:
Minor fixes for #16 comments

ezimuel commented:
@weierophinney and @webimpress I updated the code with all your comments. Thanks for reviewing it.

weierophinney commented on src/UserRepository/HtpasswdFactory.php:
```diff
@@ -17,9 +17,7 @@ class HtpasswdFactory
-        $config = $container->has('config') ? $container->get('config') : [];
-        $htpasswd = $config['authentication']['htpasswd'] ?? null;
-
+        $htpasswd = $container->get('config')['authentication']['htpasswd'] ?? null;
```
Why this particular change?

ezimuel replied:
Just because it's less line of code and we don't need `$config`.

weierophinney replied:
The previous version wouldn't fail immediately if the `config` service was not present, but, ultimately, it would fail anyways, so I'm okay with the change.

michalbundyra replied:
@ezimuel @weierophinney wait... shouldn't container throw an exception (ContainerException) when service is not defined?
Please bare in mind we are mocking container always in tests, so not sure if it's right change.

michalbundyra replied:
Ofc, check this out:

```php
    public function testContainer()
    {
        $container = new ServiceManager();
        $container->get('config');
    }
```

throws exception:
```
...Test::testContainer
Zend\ServiceManager\Exception\ServiceNotFoundException: Unable to resolve service "config" to a factory; are you certain you provided it during configuration?
```

weierophinney replied:
Right, and that's what I was getting at. The change that @ezimuel makes here will now emit that particular exception if the `config` service is not defined. It raises another exception if the service is defined, but the expected configuration is missing. That seems like a reasonable change, as it actually gets more precise about the cause of failure.

michalbundyra replied:
@weierophinney So we are missing that test case in tests now. By default container is prophesized to return null on all methods, but in that case exception should be thrown.

weierophinney replied:
Okay, will add that test and push to @ezimuel 's branch shortly.

weierophinney commented on src/UserRepository/PdoDatabase.php:
```diff
@@ -64,20 +71,32 @@ public function authenticate(string $credential, string $password = null) : ?Use
+            $stmt = $this->pdo->prepare($this->config['sql_get_roles']);
+        } catch (PDOException $e) {
+            throw new Exception\RuntimeException(sprintf(
+                "Error on sql_get_rols configuration: %s",
```
Two things on this one:

- Use single quotes (for consistency)
- I'd rephrase this to 'Error preparing retrieval of user roles: %s'; using "prepare" here helps the developer identify what sort of action was being attempted.

weierophinney commented on src/UserRepository/PdoDatabase.php:
```diff
@@ -64,20 +71,32 @@ public function authenticate(string $credential, string $password = null) : ?Use
+        if (false === $stmt) {
+            throw new Exception\RuntimeException(
+                "Error on sql_get_rols configuration"
+            );
```
Emit exactly the same exception message here as above, as this particular conditional is testing the same thing (error with statement preparation), but with PDO exceptions disabled. The only change is that the `%s` would become `unknown error`.

michalbundyra replied:
So maybe it should be moved into `Exception\RuntimeException` class with static method and parameter for `%s` as we already have in other expressive components?

weierophinney reviewed (changes requested):
Found a few more changes for you...

michalbundyra commented on docs/book/v1/user-repository.md:
````diff
@@ -42,3 +42,80 @@ result will be a `UserInterface` instance, otherwise a null value is returned.
+```php
+'config' => [
+    'authentication' => [
+        'htpasswd' => 'insert the path to htpasswd file'
````
I would add everywhere trailing commas in arrays in the docs as well (this is going to be a part of new coding-standard and it would be good to have it also in configuration, because folks could copy from there)

ezimuel committed:
Incorporated the latest comments on #16

ezimuel commented:
@weierophinney and @webimpress incorporated the latest comments. Thanks again.

weierophinney committed:
Edits to documentation

- Edited docs covering UserRepository configuration for grammar, flow, etc.
- Updated ConfigProvider docblocks to show potential values accepted.

weierophinney committed:
Test for container raising exception on config retrieval

Modifies HtpasswdFactoryTest::testInvokeWithMissingConfig to have the
container throw an exception from `get('config')` instead of returning
null, and demonstrates that this will not be caught by the factory.

Adds the same test to PdoDatabaseFactoryTest.
